GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using GitCommands;
+using Microsoft.Win32;
+
+namespace GitUI.CommandsDialogs.SettingsDialog.ShellExtension
+{
+    public static class ShellExtensionManager
+    {
+        /// <summary>
+        /// Checks if necessary for registration files were found
+        /// </summary>
+        /// <returns>false only if all necessary files were found, otherwise true</returns>
+        public static bool IsNotInstalled()
+        {
+            string path32 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx32Name);
+            string path64 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx64Name);
+            return string.IsNullOrEmpty(path32) || (Environment.Is64BitOperatingSystem && string.IsNullOrEmpty(path64));
+        }
+
+        /// <summary>
+        /// Checks if shell extension is properly registered in system
+        /// </summary>
+        /// <returns>false only if all checks are positive, otherwise true</returns>
+        public static bool IsNotRegistered()
+        {
+            return string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.LocalMachine, @"Software\Microsoft\Windows\CurrentVersion\Shell Extensions\Approved",
+                    "{3C16B20A-BA16-4156-916F-0A375ECFFE24}")) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"*\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\Background\shellex\ContextMenuHandlers\GitExtensions2",
+                    null));
+        }
+
+        /// <summary>
+        /// Register shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void RegisterShellExtension() => RegisterAllShellExtensionDlls("{0}");
```

Comment:
```suggestion
        public static void Register() => RegisterAllShellExtensionDlls("{0}");
```

Reply:
Why do we need to pass "{0}"?

Reply:
The better name for RegisterAllShellExtensionDlls is RunRegSvrForShellExtensionDlls. And parameter name is argumentsPattern, so here we should provide valid arguments pattern, which for register command is "{0}".
Before refactoring there was a code with string concatenations and strreplace so I decided to rewrite it.